packages/opentelemetry-resource-detector-aws/src/detectors/AwsEcsDetector.ts

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Detector,
+  Resource,
+  ResourceDetectionConfigWithLogger,
+  CONTAINER_RESOURCE,
+} from '@opentelemetry/resources';
+import * as util from 'util';
+import * as fs from 'fs';
+import * as os from 'os';
+
+/**
+ * The AwsEcsDetector can be used to detect if a process is running in AWS
+ * ECS and return a {@link Resource} populated with data about the ECS
+ * plugins of AWS X-Ray. Returns an empty Resource if detection fails.
+ */
+export class AwsEcsDetector implements Detector {
+  readonly CONTAINER_ID_LENGTH = 64;
+  readonly DEFAULT_CGROUP_PATH = '/proc/self/cgroup';
+  private static readFileAsync = util.promisify(fs.readFile);
+
+  async detect(config: ResourceDetectionConfigWithLogger): Promise<Resource> {
+    if (
+      !process.env.ECS_CONTAINER_METADATA_URI_V4 &&
+      !process.env.ECS_CONTAINER_METADATA_URI
+    ) {
+      config.logger.debug('AwsEcsDetector failed: Process is not on ECS');
+      return Resource.empty();
+    }
+
+    let hostName, containerId;
+    try {
```

Comment:
To avoid the use of `let`, please move the try/catch into functions that return values or undefined:

```typescript
const [hostName, containerId] = await Promise.all([
  this._getHostname(),
  this._getContainerId()
]);
```

```typescript
private async _getContainerId(): Promise<string | undefined> {
  try {
    // do all work to get the container id here and return it
  } catch (err) {
    // log error
    return;
  }
}
```

Reply:
Typically `let` is only used for variables which change many times over the course of a function like a counter or a string that is being built.

Reply:
Make sense, let me do it!